# Assignment 1
You should submit the **UniversityNumber.ipynb** file and your final prediction file **UniversityNumber.test.out** to moodle. Make sure your code does not use your local files and that the results are reproducible. Before submitting, please **run your notebook and keep all running logs** so that we can check.

## 1 $n$-gram Language Model

In [ ]:
!wget -O train.txt https://raw.githubusercontent.com/ranpox/comp7607-fall2022/main/assignments/A1/data/lm/train.txt
!wget -O dev.txt https://raw.githubusercontent.com/ranpox/comp7607-fall2022/main/assignments/A1/data/lm/dev.txt
!wget -O test.txt https://raw.githubusercontent.com/ranpox/comp7607-fall2022/main/assignments/A1/data/lm/test.txt

### 1.1 Building vocabulary

#### Code

In [13]:

MINIMAL_FREQUENCY = 3
START_TOKEN = '<s>'
END_TOKEN = '</s>'
UNKNOWN_TOKEN = '<UNK>'

FILE_PATH = 'data/lm'
TEST_FILENAME = f'{FILE_PATH}/test.txt'
DEV_FILENAME = f'{FILE_PATH}/dev.txt'
TRAIN_FILENAME = f'{FILE_PATH}/train.txt'

word_frequency = {}

def load_file(file_name):
    texts = []
    with open(file_name, encoding='utf-8') as f:
        for line in f.readlines():
            text = line.strip().lower().split(' ')
            text.insert(0, START_TOKEN)
            text.append(END_TOKEN)
            texts.append(text)
    return texts

def cal_frequency(texts):
    global word_frequency
    for text in texts:
        for word in text:
            if word not in word_frequency:
                word_frequency[word] = 1
            else:
                word_frequency[word] += 1

def add_unk():
    global word_frequency
    tmp = {}
    tmp[UNKNOWN_TOKEN] = 0
    for key in word_frequency:
        value = word_frequency[key]
        if value < MINIMAL_FREQUENCY:
            tmp[UNKNOWN_TOKEN] += value
        else:
            tmp[key] = value
    word_frequency = tmp

train_texts = load_file(TRAIN_FILENAME)
dev_texts = load_file(DEV_FILENAME)
test_texts = load_file(TEST_FILENAME)

cal_frequency(train_texts)
print(f'have {len(word_frequency)} tokens.')

add_unk()
print(f'after add unknown token, have {len(word_frequency)} tokens.')


have 46310 tokens.
after add unknown token, have 20663 tokens.


#### Discussion

There are 46310 tokens in the training dataset. After removing tokens that occur less than three times, there are 20663 tokens.
Use V as the size of vocabulary. Then there are $V^{n}$ parameters in n-gram models. Most of the parameters are zero. The number of none-zero parameters is linear to the total length of the training dataset.


### 1.2 $n$-gram Language Modeling

#### Code

In [14]:
## unigram
def unigram(train_texts):
    model = {'c_up': {}, 'c_down': {}}
    for text in train_texts:
        for word in text:
            if word not in model['c_up']:
                model['c_up'][word] = 1
            else:
                model['c_up'][word] += 1
    return model

def bigram(train_texts):
    model = {'c_up': {}, 'c_down': {}}
    
    return model

IndentationError: expected an indented block (2016616479.py, line 12)

#### Discussion

### 1.3 Smoothing

#### 1.3.1 Add-one (Laplace) smoothing

##### Code

##### Discussion

##### Optional: Add-k smoothing

###### Code

###### Discussion

#### 1.3.2 Linear Interpolation

##### Code

##### Discussion

##### Optional: Optimization

###### Discussion

## 2 Preposition Prediction

In [ ]:
!wget -O dev.in https://raw.githubusercontent.com/ranpox/comp7607-fall2022/main/assignments/A1/data/prep/dev.in
!wget -O dev.out https://raw.githubusercontent.com/ranpox/comp7607-fall2022/main/assignments/A1/data/prep/dev.out